eyeball-im-util/src/vector/sort.rs

```diff
@@ -0,0 +1,613 @@
+        + Clone,
+{
+    /// Create a new `Filter` with the given (unfiltered) initial values, stream
+    /// of `VectorDiff` updates for those values, and filter.
+    pub fn new(
+        initial_values: Vector<VectorDiffContainerStreamElement<S>>,
+        inner_stream: S,
+        compare: F,
+    ) -> (Vector<VectorDiffContainerStreamElement<S>>, Self) {
+        let mut initial_values = initial_values.into_iter().enumerate().collect::<Vector<_>>();
+        initial_values.sort_by(|(_, left), (_, right)| compare(left, right));
+
+        (
+            initial_values.clone().into_iter().map(|(_, value)| value).collect(),
+            Self {
+                inner_stream,
+                compare,
+                buffered_vector: initial_values,
+                ready_values: Default::default(),
+            },
+        )
+    }
+
+    /// For testing purposes only.
+    #[doc(hidden)]
+    pub fn buffered_vector(&self) -> &Vector<(UnsortedIndex, VectorDiffContainerStreamElement<S>)> {
+        &self.buffered_vector
+    }
+}
+
+impl<S, F> Stream for Sort<S, F>
+where
+    S: Stream,
+    S::Item: VectorDiffContainer,
+    F: Fn(&VectorDiffContainerStreamElement<S>, &VectorDiffContainerStreamElement<S>) -> Ordering
+        + Clone,
+{
+    type Item = S::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Option<Self::Item>> {
+        self.project().poll_next(cx)
+    }
+}
+
+impl<S, F> SortProj<'_, S, F>
+where
+    S: Stream,
+    S::Item: VectorDiffContainer,
+    F: Fn(&VectorDiffContainerStreamElement<S>, &VectorDiffContainerStreamElement<S>) -> Ordering
+        + Clone,
+{
+    fn poll_next(&mut self, cx: &mut task::Context<'_>) -> Poll<Option<S::Item>> {
+        loop {
+            // First off, if any values are ready, return them.
+            if let Some(value) = S::Item::pop_from_buffer(self.ready_values) {
+                return Poll::Ready(Some(value));
+            }
+
+            // Poll `VectorDiff`s from the `inner_stream`.
+            let Some(diffs) = ready!(self.inner_stream.as_mut().poll_next(cx)) else {
+                return Poll::Ready(None);
+            };
+
+            // Consume and apply the diffs if possible.
+            let ready = diffs.push_into_buffer(self.ready_values, |diff| {
+                handle_diff_and_update_buffered_vector(
+                    diff,
+                    self.compare.clone(),
+                    self.buffered_vector,
+                )
+            });
+
+            if let Some(diff) = ready {
+                return Poll::Ready(Some(diff));
+            }
+
+            // Else loop and poll the streams again.
+        }
+    }
+}
+
+// Map a `VectorDiff` to potentially `VectorDiff`s. Keep in mind that
+// `buffered_vector` contains the sorted values.
+fn handle_diff_and_update_buffered_vector<T, F>(
+    diff: VectorDiff<T>,
+    compare: F,
```

Comment:
I think if this was `compare: &F` the function would still compile, and you could remove the `.clone()` on `self.compare` above in `poll_next`. Then theh `Clone` bound on `F` could likely be removed too.